### PR TITLE
RFD 42 - S3 KMS Encryption

### PR DESCRIPTION
**Pull request by xacrimon**

### CODE REVIEW

File: rfd/0042-s3-kms-encryption.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+The configured KMS CMK needs to have a standard spec configuration and must be symmetric.
+
+Below template KMS policies are provided for restricting access to
+KMS keys and the needed permissions. The policies are to be filled in with
+the IAM user the authentication nodes authenticate as.
+
+### Encryption/Decryption
+
+This policy allows an IAM user to encrypt and decryption objects.
```
Comment (klizhentas) [inline rfd/0042-s3-kms-encryption.md:34]: and decrypt

File: rfd/0042-s3-kms-encryption.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        "kms:GenerateDataKey*",
+        "kms:DescribeKey"
+      ],
+      "Resource": "*"
+    }
+  ]
+}
+```
+
+### Encryption/Decryption with separate clusters
```
Comment (klizhentas) [inline rfd/0042-s3-kms-encryption.md:74]: Would be helpful if you added a AWS IAM policy boundary to show admins how they can create users who won't be capable of accessing encrypt/decrypt with the key as well.

File: rfd/0042-s3-kms-encryption.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+## Why
+
+Encrypting objects in S3 with a custom CMK allows for additional security as it
+can be used to restrict access to sensitive data like session recordings
+to those with read privileges to S3 buckets storing said sensitive data.
+
+## Details
+
+As per the current scheme for configuring storage backends like S3 an optional URL
+parameter will be added to the S3 storage url with the name of `sse_kms_key`.
```
Comment (klizhentas) [inline rfd/0042-s3-kms-encryption.md:22]: Can you give an example of the URL?

**Review by klizhentas — APPROVED** [review]

File: rfd/0042-s3-kms-encryption.md (hunk 1 above)
Comment (xacrimon) [reply rfd/0042-s3-kms-encryption.md]: Fixed

File: rfd/0042-s3-kms-encryption.md (hunk 3 above)
Comment (xacrimon) [reply rfd/0042-s3-kms-encryption.md:22]: Fixed

File: rfd/0042-s3-kms-encryption.md (hunk 2 above)
Comment (xacrimon) [reply rfd/0042-s3-kms-encryption.md:79]: IAM users have to be manually specified in the policy to have access so I added a section explaining that.

**Comment by xacrimon** [comment]
@russjones please review

**Comment by russjones** [comment]
@xacrimon

1. Did you test the two cluster setup (one that can only write sessions, one that can read sessions?)

2. 

> IAM users have to be manually specified in the policy to have access so I added a section explaining that.

Is this even the case if you give your IAM user `AdministratorAccess`? I think when I tried it with `AdministratorAccess` I was able to access KMS.

**Comment by xacrimon** [comment]
> @xacrimon
> 
> 1. Did you test the two cluster setup (one that can only write sessions, one that can read sessions?)
> 
> > IAM users have to be manually specified in the policy to have access so I added a section explaining that.
> 
> Is this even the case if you give your IAM user `AdministratorAccess`? I think when I tried it with `AdministratorAccess` I was able to access KMS. I wouldn't rely on the documentation here, can you test it out?

@russjones
1. Yes, and it seemed to work well. It's possible though that some minor code change I am unaware of would be required to make everything run smoothly on the read cluster to indicate that it is indeed a read only cluster for that purpose but I haven't seen anything that would indicate it.

2. It isn't, `AdministratorAccess` is close to equivalent to a root account which means you can access and administrate KMS keys. It's my understanding that having IAM users with such access is distinctly bad practice and that the IAM limited administrators feature should be used instead if you do need such administrators. That said you can use a `Deny` clause in the policy in order to explicitly deny some administrator users the KMS key admin permission in the key policy but I didn't include this for the above reason. I could add another note explaining that however.

**Comment by xacrimon** [comment]
@russjones Updated the RFD with a section on admins. Please re-review.

**Comment by russjones** [comment]
@xacrimon What parts of the the following are relevant for admins not to have?
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "kms:CreateAlias",
                "kms:CreateKey",
                "kms:DeleteAlias",
                "kms:Describe*",
                "kms:GenerateRandom",
                "kms:Get*",
                "kms:List*",
                "kms:TagResource",
                "kms:UntagResource",
                "iam:ListGroups",
                "iam:ListRoles",
                "iam:ListUsers"
            ],
            "Resource": "*"
        }
    ]
}
```

**Comment by xacrimon** [comment]
@russjones Did further testing and as it turns out IAM `AdministratorAccess` or any other IAM policy does not give you any permissions to use existing keys. It does allow you to create new keys of course but an IAM user will never be able to access a key it didn't create or wasn't explicitly granted access to in the key policy which I already have tested templates for.

Due to this I believe this should be ready, I updated the last block to clarify that no users have access by default but I don't think there is anything more to it.

**Comment by xacrimon** [comment]
@russjones review please

**Review by russjones — APPROVED** [review]

**Comment by xacrimon** [comment]
Okay, did my second verification pass of testing and I've had to remove the ability to separate the encrypt and decrypt IAM permissions. I made a mistake somewhere in my previous testing and as it turns out, AWS S3 does not for some reason allow you split up KMS encrypt and decrypt permissions for a given user when using S3 KMS SSE.

**Comment by xacrimon** [comment]
Okay, I performed a second testing and security check pass and discovered some issues. Due to limitations in how S3 KMS SSE works, this dual cluster setup will not work to provide extra security. This wasn’t super well documented but essentially S3 KMS SSE does not allow you to completely limit decryption access for an IAM user which has the permissions needed to upload. This restriction is only present for SSE as far as I am aware.

This means the Teleport Auth will always have the ability to decrypt sessions it uploads. Now, there’s a couple things we can do. We guide admins to restrict read access to sessions within Teleport to prevent everyone except high level cluster admins from reading the sessions. I’ve current removed the dual cluster policy and updated the other policy with my findings to restrict access.

So you can still prevent AWS admins from using the key and viewing the sessions but the Teleport Auth IAM user will always have decryption access. 

The dual cluster configuration could still be employed but would not prevent root level admins of the main cluster from reading the sessions.

Is the dual cluster confit something we still want in that case? If so I’ll add a section on how to set it up but the main cluster would still use the same key policy permissions as the decryption cluster.

**Comment by xacrimon** [comment]
cc @russjones

**Review by russjones — APPROVED** [review]

**Pull request merged**